chore(lint, workspace): Orderliness and break avoiding lint rules, package.json flags

Krisztiaan opened the pull request:
`import/order`: group and order imports to make imports easily scannable and deterministic
`no-restricted-imports`: warn about using `src/*` because typescript does not replace these paths so the `import`s will be any in the `d.ts` files

PARTIALLY Closes #1297

peterp commented on packages/eslint-config/index.js:
```diff
@@ -124,5 +124,32 @@ module.exports = {
+        },
+      },
+    ],
+    'no-restricted-imports': [
+      'warn',
+      {
+        patterns: [
+          'src/*', // Typescript doesn't change import paths, which leads to broken .d.ts files without bundling. Use relative.
+        ],
+      },
+    ],
```
Are you're suggesting that people do not use the `src` alias that we provide? Could you expand on the comment you've made here, maybe with an example?

Krisztiaan replied:
Look at `packages/api/dist/auth/index.d.ts` for example. `L29` the variable `context` is any, because `GlobalContext` can't be resolved by typescript without the knowledge of the `tsconfig` path mapping. The problem is that `tsc` does interpret `src/foo` as valid for typechecking, but when generating output, it'll actually not change the path, so there'll be no way for the user's `tsc` to know where that `src/foo` actually points to.

peterp replied:
🤦 Oh wow, that's pretty lame.

Ok! We share this configuration with consumer's of Redwood. Could you add this rule in `./package.json` instead of here?

Since consumers of Redwood are not going to be distributing typeDefs it should be OK for them to use the `src` alias.

Krisztiaan replied:
Oh shoot, got it, didn't pay much attention to why this eslint config is not just an `.eslintrc` 😅 
I'll move it in a bit.

Krisztiaan replied:
Done

peterp replied:
Thanks!

Krisztiaan commented:
A version of [`tspath`](https://github.com/duffman/tspath) could also be used to transform `src/*` into something usable.

peterp commented on package.json:
```diff
@@ -39,7 +41,42 @@
+    "extends": "@redwoodjs/eslint-config",
+    "rules": {
+      "import/order": [
+        "error",
```
Could we make this a warning instead, so that editors should still auto-format it for people that have the correct extensions installed, but it won't block us from merging a PR if it's not 100% adherence to the standard.

peterp commented on packages/auth/src/authClients/index.ts:
```diff
@@ -1,18 +1,18 @@
 import type { MagicLink, MagicUser } from './magicLink'
-import type { Firebase } from './firebase'
+import { magicLink } from './magicLink'
+import type { NetlifyIdentity } from './netlify'
```
Hmmm... this is unfortunate. Is there a way to specify that "type imports" should be right at the top of the file?

Krisztiaan replied:
Probably, I'll look into it

peterp replied:
No luck with this?

Krisztiaan replied:
Unfortunately, no. See: https://github.com/benmosher/eslint-plugin-import/issues/645

peterp replied:
Ok, no worries. It's better than nothing.

peterp commented:
> Something like typescript-transform-paths could also be used to avoid needing the src/* rule though.

Oooh, that's really cool - I really like the `src/` alias, could we try this out?

Krisztiaan commented:
`src` is success, made some tuning on the import order, and extracted the eslint config from `package.json`. Moved scripts in `package.json` to start of file, as it's one of the most important things to overview when opening that file.

Added `ttsc` (not a typo!) that provides plugin capabilities to the typescript compiler. Needed for the `src/*` mapping.

peterp commented on .vscode/settings.json:
```diff
@@ -9,11 +9,22 @@
     "statusBarItem.hoverBackground": "#ce7350",
     "statusBar.foreground": "#e7e7e7",
-    "statusBar.border": "#b85833"
+    "activityBar.activeBackground": "#ce7350",
+    "activityBar.activeBorder": "#82dc9b",
+    "activityBar.background": "#ce7350",
+    "activityBar.foreground": "#15202b",
+    "activityBar.inactiveForeground": "#15202b99",
+    "activityBarBadge.background": "#82dc9b",
+    "activityBarBadge.foreground": "#15202b",
+    "titleBar.activeBackground": "#b85833",
+    "titleBar.activeForeground": "#e7e7e7",
+    "titleBar.inactiveBackground": "#b8583399",
+    "titleBar.inactiveForeground": "#e7e7e799"
```
Was this from peacock?

Krisztiaan replied:
I need to find out how to disable this behavior. Yes, it is from peacock

peterp replied:
I think if it detects that you're setting the status bar colour it takes over and says: "OK PEACOCK TIME!"

peterp commented:
Hey @Krisztiaan, I would love to get this merged, I've finally reviewed everything!

![image](https://user-images.githubusercontent.com/44849/100155397-7eff2a80-2ea7-11eb-84e8-86df94bea14a.png)

I tried to fix the merge conflicts, but I can't seem to get the tests to complete on my end. They're stalling infinitely.

peterp commented:
Ah, I see the same tests are failing in the previous tests.

peterp commented:
Damnit, I would love to merge this.

peterp commented on package.json:
```diff
@@ -1,6 +1,16 @@
 {
+  "scripts": {
+    "build": "lerna run build:js && ttsc --build --verbose",
+    "build:clean": "rimraf ./packages/**/dist",
+    "build:watch": "ttsc --build && lerna run build:watch --parallel",
+    "test": "lerna run test --stream -- --colors --maxWorkers=4",
+    "lint": "eslint packages",
+    "lint:fix": "eslint --fix packages"
+  },
```
I actually leave these at the bottom because I run `cat package.json` to see what I can run.

peterp replied:
And I guess you're doing this for the exact reason, but you're opening it in an editor and reading from the top.

peterp commented on package.json:
```diff
@@ -25,28 +34,21 @@
     "msw": "0.21.2",
     "nodemon": "^2.0.6",
     "rimraf": "^3.0.2",
+    "ts-node": "^9.0.0",
```
Do we use ts-node?

Krisztiaan replied:
No, good catch

peterp reviewed (approved):
LGTM!

AntonioMeireles commented:
**late to the party, (sorry)** - is there any good reason for the rules that were inside `.eslintrc` not to be in plain stock `eslint-config` so that rw projects would also consume them out of the box ?

Krisztiaan commented:
@AntonioMeireles probably me misunderstanding [this](https://github.com/redwoodjs/redwood/pull/1381#discussion_r510889621)

AntonioMeireles commented:
@Krisztiaan 

actually you didn't 😄  - at the time it made sense. it's just that the `'no-restricted-imports'`that was the culprit ended going away and the remaining bits imho are idiomatic and generic enough that they could just be shipped in `eslint-config`.

Krisztiaan commented:
That's what I mean by misunderstanding, I moved everything to project level instead of just the `src` rule
